From b3f200dabe8085b130c848c7ff7dca355afc8881 Mon Sep 17 00:00:00 2001
From: Sammy He <r62914@freescale.com>
Date: Wed, 17 Dec 2008 03:23:19 +0000
Subject: [PATCH] ENGR00101947 optimize vpu clock gate between StartOneFrame and GetOutputInfo

Clock off after decoding/encoding interrupt received immediately,
then clock on again when calling GetOutputInfo.

Signed-off-by: Sammy He <r62914@freescale.com>
---
 drivers/mxc/vpu/mxc_vpu.c |   15 +++++++++++++--
 1 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/drivers/mxc/vpu/mxc_vpu.c b/drivers/mxc/vpu/mxc_vpu.c
index bbe1234..196260c 100644
--- a/drivers/mxc/vpu/mxc_vpu.c
+++ b/drivers/mxc/vpu/mxc_vpu.c
@@ -143,6 +143,15 @@ static irqreturn_t vpu_irq_handler(int irq, void *dev_id)
 	if (dev->async_queue)
 		kill_fasync(&dev->async_queue, SIGIO, POLL_IN);
 
+	/*
+	 * Clock is gated on when dec/enc started, gate it off when
+	 * interrupt is received.
+	 */
+	spin_lock(&vpu_lock);
+	if (clkgate_refcount > 0 && !(--clkgate_refcount))
+		clk_disable(vpu_clk);
+	spin_unlock(&vpu_lock);
+
 	codec_done = 1;
 	wake_up_interruptible(&vpu_queue);
 
@@ -277,10 +286,12 @@ static int vpu_ioctl(struct inode *inode, struct file *filp, u_int cmd,
 	case VPU_IOC_CLKGATE_SETTING:
 		{
 			u32 clkgate_en;
+			unsigned long lock_flags;
+
 			if (get_user(clkgate_en, (u32 __user *) arg))
 				return -EFAULT;
 
-			spin_lock(&vpu_lock);
+			spin_lock_irqsave(&vpu_lock, lock_flags);
 			if (clkgate_en) {
 				if (++clkgate_refcount == 1)
 					clk_enable(vpu_clk);
@@ -289,7 +300,7 @@ static int vpu_ioctl(struct inode *inode, struct file *filp, u_int cmd,
 				    && !(--clkgate_refcount))
 					clk_disable(vpu_clk);
 			}
-			spin_unlock(&vpu_lock);
+			spin_unlock_irqrestore(&vpu_lock, lock_flags);
 
 			break;
 		}
-- 
1.5.4.4

